eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -511,16 +492,16 @@ object Helpers {
       }
       val (closingTx, closingSigned) = makeClosingTx(keyManager, commitments, localScriptPubkey, remoteScriptPubkey, remoteClosingFee)
       val signedClosingTx = Transactions.addSigs(closingTx, keyManager.fundingPublicKey(commitments.localParams.fundingKeyPath).publicKey, remoteParams.fundingPubKey, closingSigned.signature, remoteClosingSig)
-      Transactions.checkSpendable(signedClosingTx).map(x => signedClosingTx.tx).recover { case _ => throw InvalidCloseSignature(commitments.channelId, signedClosingTx.tx) }
+      Transactions.checkSpendable(signedClosingTx).map(_ => signedClosingTx.tx).recover { case _ => throw InvalidCloseSignature(commitments.channelId, signedClosingTx.tx) }
     }
 
-    def generateTx(desc: String)(attempt: Try[TransactionWithInputInfo])(implicit log: LoggingAdapter): Option[TransactionWithInputInfo] = {
+    def generateTx(desc: String)(attempt: Try[Either[TxGenerationSkipped, TransactionWithInputInfo]])(implicit log: LoggingAdapter): Option[TransactionWithInputInfo] = {
```

Comment:
I guess the main reason why tx generation was throwing an exception is because of this function.
We really really want to make sure our tx generation doesn't fail with an unhandled exception, so it kind of makes sense to wrap all of them in a `Try` like we currently do.

But what exactly may throw here? The tx signing? If that's the case we should probably go deeper and mark those functions as return a `Try`?

Reply:
> But what exactly may throw here? The tx signing? If that's the case we should probably go deeper and mark those functions as return a Try?

I think this is the right direction but we won't be able to completely get rid of the `Try`  until we address all the possible reasons for why something might throw an exception down there. If we can catch in most of the errors (and translate them to `TxGenerationSkipped`) then the signing part should be the only one left to be done inside the `Try`. This is also a bit dangerous because if we overlook some code that can still throw exception we might have a failing tx generation because of an unhandled exception.

Reply:
> until we address all the possible reasons for why something might throw an exception down there.

That's exactly the goal ;)
We are fortunate enough to be using a functional language, where exceptions shouldn't be hidden below us. We should be able to tell whether some code we're calling is pure or not, and may throw exceptions or not. Otherwise we're not doing better than JS...

I don't think it's acceptable for code that handles users' money to say "I'm not exactly sure what this code does so I'll wrap it in a `Try` to be safe". I think we should know exactly what code may throw an exception that we want to catch, and use the type system to provide accurate function signatures. Maybe I'm too optimistic, but if we could get there our codebase would feel very very safe.

Reply:
I agree on the principle and i think we can leverage the type system to do better. However scala is not a pure functional language and it still has ugly things like `null`s and NPE (which are absent in functional programming), so for example [here](https://github.com/ACINQ/eclair/pull/1366/files#diff-aeb98076f262543b70a86753bd3719adR519) if we pass a `null` commitTx parameter it will throw a NPE in the row below and if not caught by an external `Try..match` this can be fatal. 
Having said that i think we **can** work out all the possible exceptions that can be thrown by our critical code and wrap them into a `TxGenerationSkipped`, as you said you probably we still need to wrap the signing methods around a `Try` but that should be all of the exception-catching.

Reply:
> scala is not a pure functional language

True, maybe we can only get this far with Scala...when Haskell??

Reply:
We are also using Akka, with its "Let it crash" philosophy. In general we do not even try to catch exceptions (which is a lost cause as @araspitzu explained). We do have a [`handleExceptions`](https://github.com/ACINQ/eclair/blob/master/eclair-core/src/main/scala/fr/acinq/eclair/channel/Channel.scala#L2254-L2261) in Channels.scala, but it's merely to react to the inevitable exceptions, not to proactively handle them.

In this particular piece of code, which is very sensitive, we have been extra-cautious and have been using `Try`s as an added layer of defense against unexpected behavior. The reason is mainly that the underlying signature code contains JNI calls to libsecp256k1.

I would be fine with having `generateTx` return an `Either`, provided that exceptions are catched within the method, and converted to a `Left`.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1366/commits/a816c2fac0299f36df4964461f2b8ed10b611b30, I think it's nice to push this `Try` inside `generateTx` instead of repeating it every time we call `generateTx`.